File: service/src/test/java/org/coner/core/util/ApiRequestTestUtils.java
package org.coner.core.util;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Date;
import java.util.Set;

import org.coner.core.api.request.AddCompetitionGroupRequest;
import org.coner.core.api.request.AddCompetitionGroupSetRequest;
import org.coner.core.api.request.AddEventRequest;
import org.coner.core.api.request.AddHandicapGroupRequest;
import org.coner.core.api.request.AddHandicapGroupSetRequest;
import org.coner.core.api.request.AddRegistrationRequest;
import org.coner.core.api.request.AddRunRequest;

import com.google.common.collect.Sets;

public final class ApiRequestTestUtils {

    private ApiRequestTestUtils() {

    }

    public static AddEventRequest fullAddEvent() {
        return fullAddEvent(
                TestConstants.EVENT_NAME,
                TestConstants.EVENT_DATE,
                TestConstants.HANDICAP_GROUP_SET_ID,
                TestConstants.COMPETITION_GROUP_SET_ID,
                TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION
        );
    }

    public static AddEventRequest fullAddEvent(
            String name,
            Date date,
            String handicapGroupSetId,
            String competitionGroupSetId,
            int maxRunsPerRegistration
    ) {
        AddEventRequest request = new AddEventRequest();
        request.setName(name);
        request.setDate(date);
        request.setHandicapGroupSetId(handicapGroupSetId);
        request.setCompetitionGroupSetId(competitionGroupSetId);
        request.setMaxRunsPerRegistration(maxRunsPerRegistration);
        return request;
    }

    public static AddCompetitionGroupRequest fullAddCompetitionGroup() {
        return fullAddCompetitionGroup(
                TestConstants.COMPETITION_GROUP_NAME,
                TestConstants.COMPETITION_GROUP_GROUPING,
                TestConstants.COMPETITION_GROUP_FACTOR,
                TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.toString()
        );
    }

    public static AddCompetitionGroupRequest fullAddCompetitionGroup(
            String name,
            boolean grouping,
            BigDecimal handicapFactor,
            String resultTimeType
    ) {
        AddCompetitionGroupRequest addCompetitionGroupRequest = new AddCompetitionGroupRequest();
        addCompetitionGroupRequest.setName(name);
        addCompetitionGroupRequest.setGrouping(grouping);
        addCompetitionGroupRequest.setFactor(handicapFactor);
        addCompetitionGroupRequest.setResultTimeType(resultTimeType);
        return addCompetitionGroupRequest;
    }

    public static AddCompetitionGroupSetRequest fullAddCompetitionGroupSet() {
        return fullAddCompetitionGroupSet(
                TestConstants.COMPETITION_GROUP_SET_NAME,
                Sets.newHashSet(TestConstants.COMPETITION_GROUP_ID)
        );
    }

    public static AddCompetitionGroupSetRequest fullAddCompetitionGroupSet(
            String competitionGroupSetName,
            Set<String> competitionGroupSetIds
    ) {
        AddCompetitionGroupSetRequest addCompetitionGroupSetRequest = new AddCompetitionGroupSetRequest();
        addCompetitionGroupSetRequest.setName(competitionGroupSetName);
        addCompetitionGroupSetRequest.setCompetitionGroupIds(competitionGroupSetIds);
        return addCompetitionGroupSetRequest;
    }

    public static AddHandicapGroupRequest fullAddHandicapGroup() {
        return fullAddHandicapGroup(
                TestConstants.HANDICAP_GROUP_NAME,
                TestConstants.HANDICAP_GROUP_FACTOR
        );
    }

    public static AddHandicapGroupRequest fullAddHandicapGroup(
            String name,
            BigDecimal factor
    ) {
        AddHandicapGroupRequest addHandicapGroupRequest = new AddHandicapGroupRequest();
        addHandicapGroupRequest.setName(name);
        addHandicapGroupRequest.setFactor(factor);
        return addHandicapGroupRequest;
    }

    public static AddHandicapGroupSetRequest fullAddHandicapGroupSet(
            String name,
            Set<String> handicapGroupIds
    ) {
        AddHandicapGroupSetRequest addHandicapGroupSetRequest = new AddHandicapGroupSetRequest();
        addHandicapGroupSetRequest.setName(name);
        addHandicapGroupSetRequest.setHandicapGroupIds(handicapGroupIds);
        return addHandicapGroupSetRequest;
    }

    public static AddHandicapGroupSetRequest fullAddHandicapGroupSet() {
        return fullAddHandicapGroupSet(
                TestConstants.HANDICAP_GROUP_SET_NAME,
                Sets.newHashSet(TestConstants.HANDICAP_GROUP_ID)
        );
    }

    public static AddRegistrationRequest fullAddRegistration() {
        return fullAddRegistration(
                fullAddPerson(),
                TestConstants.HANDICAP_GROUP_ID,
                TestConstants.COMPETITION_GROUP_ID,
                TestConstants.REGISTRATION_NUMBER,
                TestConstants.REGISTRATION_CHECKED_IN
        );
    }

    public static AddRegistrationRequest fullAddRegistration(
            AddRegistrationRequest.AddPerson person, String handicapGroupId,
            String competitionGroupId,
            String number,
            boolean checkedIn) {
        AddRegistrationRequest addRegistrationRequest = new AddRegistrationRequest();
        addRegistrationRequest.setPerson(person);
        addRegistrationRequest.setHandicapGroupId(handicapGroupId);
        addRegistrationRequest.setCompetitionGroupId(competitionGroupId);
        addRegistrationRequest.setNumber(number);
        addRegistrationRequest.setCheckedIn(checkedIn);
        return addRegistrationRequest;
    }

    public static AddRegistrationRequest.AddPerson fullAddPerson() {
        return fullAddPerson(
                TestConstants.PERSON_FIRST_NAME,
                TestConstants.PERSON_MIDDLE_NAME,
                TestConstants.PERSON_LAST_NAME
        );
    }

    public static AddRegistrationRequest.AddPerson fullAddPerson(
            String firstName,
            String middleName,
            String lastName
    ) {
        AddRegistrationRequest.AddPerson addPerson = new AddRegistrationRequest.AddPerson();
        addPerson.setFirstName(firstName);
        addPerson.setMiddleName(middleName);
        addPerson.setLastName(lastName);
        return addPerson;
    }

    public static AddRunRequest fullAddRun() {
        return fullAddRun(
                TestConstants.REGISTRATION_ID,
                TestConstants.RUN_TIMESTAMP,
                TestConstants.RUN_RAW_TIME,
                TestConstants.RUN_CONES,
                TestConstants.RUN_PENALTY,
                TestConstants.RUN_RERUN,
                TestConstants.RUN_COMPETITIVE
        );
    }

    public static AddRunRequest fullAddRun(
            String registrationId,
            Instant timestamp,
            BigDecimal rawTime,
            int cones,
            String penalty,
            boolean rerun,
            boolean competitive
    ) {
        AddRunRequest addRunRequest = new AddRunRequest();
        addRunRequest.setRegistrationId(registrationId);
        addRunRequest.setTimestamp(timestamp);
        addRunRequest.setRawTime(rawTime);
        addRunRequest.setCones(cones);
        addRunRequest.setPenalty(penalty);
        addRunRequest.setRerun(rerun);
        addRunRequest.setCompetitive(competitive);
        return addRunRequest;
    }
}

File: service/src/test/java/org/coner/core/util/ApiRequestTestUtils.kt
package org.coner.core.util

import com.google.common.collect.Sets
import org.coner.core.api.request.*
import java.math.BigDecimal
import java.time.Instant
import java.time.Year
import java.util.*

object ApiRequestTestUtils {

    @JvmOverloads
    @JvmStatic
    fun fullAddEvent(
            name: String = TestConstants.EVENT_NAME,
            date: Date = TestConstants.EVENT_DATE,
            handicapGroupSetId: String = TestConstants.HANDICAP_GROUP_SET_ID,
            competitionGroupSetId: String = TestConstants.COMPETITION_GROUP_SET_ID,
            maxRunsPerRegistration: Int = TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION
    ): AddEventRequest {
        val request = AddEventRequest()
        request.name = name
        request.date = date
        request.handicapGroupSetId = handicapGroupSetId
        request.competitionGroupSetId = competitionGroupSetId
        request.maxRunsPerRegistration = maxRunsPerRegistration
        return request
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddCompetitionGroup(
            name: String = TestConstants.COMPETITION_GROUP_NAME,
            grouping: Boolean = TestConstants.COMPETITION_GROUP_GROUPING,
            handicapFactor: BigDecimal = TestConstants.COMPETITION_GROUP_FACTOR,
            resultTimeType: String = TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.toString()
    ): AddCompetitionGroupRequest {
        val addCompetitionGroupRequest = AddCompetitionGroupRequest()
        addCompetitionGroupRequest.name = name
        addCompetitionGroupRequest.grouping = grouping
        addCompetitionGroupRequest.factor = handicapFactor
        addCompetitionGroupRequest.resultTimeType = resultTimeType
        return addCompetitionGroupRequest
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddCompetitionGroupSet(
            competitionGroupSetName: String = TestConstants.COMPETITION_GROUP_SET_NAME,
            competitionGroupSetIds: Set<String> = Sets.newHashSet(TestConstants.COMPETITION_GROUP_ID)
    ): AddCompetitionGroupSetRequest {
        val addCompetitionGroupSetRequest = AddCompetitionGroupSetRequest()
        addCompetitionGroupSetRequest.name = competitionGroupSetName
        addCompetitionGroupSetRequest.competitionGroupIds = competitionGroupSetIds
        return addCompetitionGroupSetRequest
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddHandicapGroup(
            name: String = TestConstants.HANDICAP_GROUP_NAME,
            factor: BigDecimal = TestConstants.HANDICAP_GROUP_FACTOR
    ): AddHandicapGroupRequest {
        val addHandicapGroupRequest = AddHandicapGroupRequest()
        addHandicapGroupRequest.name = name
        addHandicapGroupRequest.factor = factor
        return addHandicapGroupRequest
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddHandicapGroupSet(
            name: String = TestConstants.HANDICAP_GROUP_SET_NAME,
            handicapGroupIds: Set<String> = Sets.newHashSet(TestConstants.HANDICAP_GROUP_ID)
    ): AddHandicapGroupSetRequest {
        val addHandicapGroupSetRequest = AddHandicapGroupSetRequest()
        addHandicapGroupSetRequest.name = name
        addHandicapGroupSetRequest.handicapGroupIds = handicapGroupIds
        return addHandicapGroupSetRequest
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddRegistration(
            person: AddRegistrationRequest.AddPerson = fullAddPerson(), car: AddRegistrationRequest.AddCar = fullAddCar(), handicapGroupId: String = TestConstants.HANDICAP_GROUP_ID,
            competitionGroupId: String = TestConstants.COMPETITION_GROUP_ID,
            number: String = TestConstants.REGISTRATION_NUMBER,
            checkedIn: Boolean = TestConstants.REGISTRATION_CHECKED_IN
    ): AddRegistrationRequest {
        val addRegistrationRequest = AddRegistrationRequest()
        addRegistrationRequest.person = person
        addRegistrationRequest.car = car
        addRegistrationRequest.handicapGroupId = handicapGroupId
        addRegistrationRequest.competitionGroupId = competitionGroupId
        addRegistrationRequest.number = number
        addRegistrationRequest.isCheckedIn = checkedIn
        return addRegistrationRequest
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddPerson(
            firstName: String = TestConstants.PERSON_FIRST_NAME,
            middleName: String = TestConstants.PERSON_MIDDLE_NAME,
            lastName: String = TestConstants.PERSON_LAST_NAME
    ): AddRegistrationRequest.AddPerson {
        val addPerson = AddRegistrationRequest.AddPerson()
        addPerson.firstName = firstName
        addPerson.middleName = middleName
        addPerson.lastName = lastName
        return addPerson
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddCar(
            year: Year = TestConstants.CAR_YEAR,
            make: String = TestConstants.CAR_MAKE,
            model: String = TestConstants.CAR_MODEL,
            trim: String = TestConstants.CAR_TRIM,
            color: String = TestConstants.CAR_COLOR
    ): AddRegistrationRequest.AddCar {
        val addCar = AddRegistrationRequest.AddCar()
        addCar.year = year
        addCar.make = make
        addCar.model = model
        addCar.trim = trim
        addCar.color = color
        return addCar
    }

    @JvmOverloads
    @JvmStatic
    fun fullAddRun(
            registrationId: String = TestConstants.REGISTRATION_ID,
            timestamp: Instant = TestConstants.RUN_TIMESTAMP,
            rawTime: BigDecimal = TestConstants.RUN_RAW_TIME,
            cones: Int = TestConstants.RUN_CONES,
            didNotFinish: Boolean = TestConstants.RUN_DID_NOT_FINISH,
            disqualified: Boolean = TestConstants.RUN_DISQUALIFIED,
            rerun: Boolean = TestConstants.RUN_RERUN,
            competitive: Boolean = TestConstants.RUN_COMPETITIVE
    ) = AddRunRequest(
            registrationId = registrationId,
            timestamp = timestamp,
            rawTime = rawTime,
            cones = cones,
            isDidNotFinish = didNotFinish,
            isDisqualified = disqualified,
            isRerun = rerun,
            isCompetitive = competitive
    )
}
